cpp/src/io/json/write_json.cu

```diff
@@ -254,38 +183,208 @@ std::unique_ptr<column> struct_to_strings(table_view const& strings_columns,
-
-  // create resulting null mask
-  auto [null_mask, null_count] = cudf::detail::valid_if(
+  // scatter row_prefix, row_suffix, column_name:, value, value_separator as string_views
+  thrust::for_each(
+    rmm::exec_policy(stream),
     thrust::make_counting_iterator<size_type>(0),
-    thrust::make_counting_iterator<size_type>(strings_count),
-    [d_table = *d_table, d_narep] __device__(size_type idx) {
-      if (d_narep.is_valid()) return true;
-      return !thrust::any_of(
-        thrust::seq, d_table.begin(), d_table.end(), [idx](auto col) { return col.is_null(idx); });
-    },
-    stream,
-    mr);
-
-  return make_strings_column(strings_count,
-                             std::move(children.first),
-                             std::move(children.second),
-                             null_count,
-                             std::move(null_mask));
+    thrust::make_counting_iterator<size_type>(total_rows),
+    [tbl       = *tbl_device_view,
+     col_names = *d_column_names,
+     row_prefix,
+     row_suffix,
+     value_separator,
+     narep = narep.value(stream),
+     include_nulls,
+     d_strviews = d_strviews.begin()] __device__(auto idx) {
+      auto const row        = idx / tbl.num_columns();
+      auto const col        = idx % tbl.num_columns();
+      auto const d_str_null = tbl.column(col).is_null(row);
+      auto const this_index = row * 3 * tbl.num_columns() + 1 * row + col * 3 + 1;
+      // prefix
+      if (col == 0) {
+        d_strviews[this_index - 1] = str_pair(row_prefix.data(), row_prefix.size_bytes());
+      }
+      if (!include_nulls && d_str_null) {
+        if (col != 0) d_strviews[this_index - 1] = str_pair(nullptr, 0);
+        d_strviews[this_index]     = str_pair(nullptr, 0);
+        d_strviews[this_index + 1] = str_pair(nullptr, 0);
+      } else {
+        // if previous column was null, then we skip the value separator
+        if (col != 0)
+          if (tbl.column(col - 1).is_null(row) && !include_nulls)
+            d_strviews[this_index - 1] = str_pair(nullptr, 0);
+          else
+            d_strviews[this_index - 1] =
+              str_pair(value_separator.data(), value_separator.size_bytes());
+        auto const d_col_name = col_names.element<string_view>(col);
+        auto const d_str = d_str_null ? narep : tbl.column(col).template element<string_view>(row);
+        // column_name: value
+        d_strviews[this_index]     = str_pair(d_col_name.data(), d_col_name.size_bytes());
+        d_strviews[this_index + 1] = str_pair(d_str.data(), d_str.size_bytes());
+      }
+      // suffix
+      if (col == tbl.num_columns() - 1) {
+        d_strviews[this_index + 2] = str_pair(row_suffix.data(), row_suffix.size_bytes());
+      }
+    });
+  auto joined_col = make_strings_column(d_strviews, stream, mr);
+
+  // gather from offset and create a new string column
+  rmm::device_uvector<size_type> d_strview_offsets(strings_columns.num_rows() + 1, stream);
```

Comment:
It seems `d_strview_offsets` can be created via `make_counting_transform_iterator` thus we can avoid materialize the data:
```cpp
auto const d_strview_offsets =
  make_counting_transform_iterator(0, [](auto i) { return i * num_str_views_per_row; });
```

Reply:
I replaced `cudf::detail::gather` with `thrust::gather` after using iterator as suggested.